How To - App Service to Azure OpenAI Auth (#40568)
dotnet/docs: docs/ai/how-to/app-service-entra-auth.md

---
title: "Authenticate and Authorize App Service to Azure OpenAI using Microsoft Entra and the Semantic Kernel SDK"
description: "Learn how to authenticate and authorize your app service application to an Azure OpenAI resource using Microsoft Entra managed identities and the Semantic Kernel SDK for .NET."
author: haywoodsloan
ms.topic: how-to
ms.date: 04/19/2024
zone_pivot_groups: azure-development-tool-set-two

#customer intent: As a .NET developer, I want authenticate and authorize my App Service to Azure OpenAI using Microsoft Entra so that I can securely use AI in my .NET application.

---

> # Authenticate and Authorize App Service to Azure OpenAI using Microsoft Entra and the Semantic Kernel SDK

Reviewer: ```suggestion
# Authenticate and authorize App Service to Azure OpenAI using Microsoft Entra and the Semantic Kernel SDK
```

This article demonstrates how to use Microsoft Entra managed identities to authenticate and authorize an App Service application to an Azure OpenAI resource.

> Using a managed identity from Microsoft Entra allows your App Service application to easily access a protected Azure OpenAI resources without having to provision or rotate any secrets.

Author: It's the reader that doesn't have to manage secrets anymore. Added "manually" to clarify that.

> Using a managed identity from Microsoft Entra allows your App Service application to easily access a protected Azure OpenAI resources without having to manually provision or rotate any secrets.

Reviewer: ```suggestion
By using a managed identity from Microsoft Entra, your App Service application can easily access protected Azure OpenAI resources without having to manually provision or rotate any secrets.
```

The [Semantic Kernel SDK](/semantic-kernel/overview) enables you to easily implement Microsoft Entra authentication in your .NET application.

## Prerequisites

* An Azure account with an active subscription. [Create an account for free](https://azure.microsoft.com/free/?WT.mc_id=A261C142F).
* [.NET SDK](https://dotnet.microsoft.com/download/visual-studio-sdks)
* [`Microsoft.SemanticKernel` NuGet package](https://www.nuget.org/packages/Microsoft.SemanticKernel)
* [`Azure.Identity` NuGet package](https://www.nuget.org/packages/Azure.Identity)
* [Create and deploy an Azure OpenAI Service resource](/azure/ai-services/openai/how-to/create-resource)
* [Create and deploy a .NET application to App Service](/azure/app-service/quickstart-dotnetcore)

## Add a managed identity to App Service

### Add a system-assigned identity

:::zone target="docs" pivot="azure-portal"

1. Navigate to your app's page on the [Azure Portal](https://aka.ms/azureportal), scroll down to the **Settings** group.
1. Select **Identity**.

> 1. Within the **System assigned** tab, switch **Status** to On. Click **Save**.

Author: Yeah, it's a toggle switch where the labels are clickable: https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#add-a-system-assigned-identity.

Thanks for the guidance on UI elements, based on that _Status_ should be italicized because it's a UI text but not clicked. And **On** should be bold because it's actually clickable.

I've also updated the formatting on a few other UI elements based on this info.

:::zone-end

:::zone target="docs" pivot="azure-cli"

> Run the az webapp identity assign command to create a system-assigned identity:

Author: Since that's an actual command I put in in an inline code segment instead.

```azurecli
az webapp identity assign --name <app-name> --resource-group <group-name>
```

:::zone-end

### Add a user-assigned identity

> To create an app with a user-assigned identity, you first create the identity and then you add its resource identifier to your app config.

Reviewer: ```suggestion
To create an app with a user-assigned identity, first create the identity and then add its resource identifier to your app config.
```

:::zone target="docs" pivot="azure-portal"

1. Create a user-assigned managed identity resource according to these [instructions](/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal#create-a-user-assigned-managed-identity).

> 1. In the left navigation for your app's page, scroll down to the *Settings* group.

Reviewer: ```suggestion
1. In the left navigation for your app's page, scroll down to the **Settings** group.
```

1. Select **Identity**.
1. Select **User assigned > Add**.
1. Search for the identity you created earlier, select it, and select **Add**.

>     > [!NOTE]

Author: It's listed outside the original instruction set I pulled this info from: https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#add-a-system-assigned-identity.

Since the goal of these steps is to add the user-assigned identity, the restart is just a side-effect. So I included this as a note since it's important info to be aware of but not required for the intended goal of these instructions.

I could see making this a warning instead, because my intention with putting this into a callout is to make sure it draws attention.

    > Once you select **Add**, the app restarts.

:::zone-end

:::zone target="docs" pivot="azure-cli"

1. Create a user-assigned identity:

    ```azurecli
    az identity create --resource-group <group-name> --name <identity-name>
    ```

1. Assign the identity to your app:

    ```azurecli
    az webapp identity assign --resource-group <group-name> --name <app-name> --identities <identity-id>
    ```

:::zone-end

## Add an Azure OpenAI user role to your managed identity

:::zone target="docs" pivot="azure-portal"

1. In the [Azure Portal](https://aka.ms/azureportal), navigate to the scope you want to grant **Azure OpenAI** access to.

>     1. This can be a **Management group**, **Subscription**, **Resource group**, or a specific **Azure OpenAI** resource.

Author: I just had it as an ordered list because the parent item was in an ordered list. I updated this sub-bullet to be a unordered list instead.

@gewarren in my local preview the formatting is a little off when I mix ordered and unordered list. Does it look alright in the Learn review link?

>     1. This can be a **Management group**, **Subscription**, **Resource group**, or a specific **Azure OpenAI** resource.

Reviewer: It looks like this:

<img width="453" alt="image" src="https://github.com/dotnet/docs/assets/24882762/78385d5d-5c9b-45f8-b619-820bfc4ec720">

>     1. This can be a **Management group**, **Subscription**, **Resource group**, or a specific **Azure OpenAI** resource.

Author: That looks much better than my preview, thanks @gewarren!

>     * This can be a **Management group**, **Subscription**, **Resource group**, or a specific **Azure OpenAI** resource.

Reviewer: ```suggestion
    * The scope can be a **Management group**, **Subscription**, or **Resource group**, or a specific **Azure OpenAI** resource.
```

1. Select **Access control (IAM)** on the left navigation pane.
1. Select **Add**, then select **Add role assignment**.
1. On the **Role** tab on the next screen, select the **Cognitive Services OpenAI User** role.

> 1. On the **Members** tab, select the previously created managed identity.

Reviewer 2: ```suggestion
1. On the **Members** tab, select the managed identity.
```

> 1. On the **Review + assign** tab, select Review + assign to assign the role.

Reviewer 2: ```suggestion
1. On the **Review + assign** tab, select **Review + assign** to assign the role.
```

:::zone-end

:::zone target="docs" pivot="azure-cli"

**Resource scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{providerName}/{resourceType}/{resourceSubType}/{resourceName}"
```

**Resource group scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}"
```

**Subscription scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}"
```

**Management group scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/providers/Microsoft.Management/managementGroups/{managementGroupName}"
```

:::zone-end

## Implement token-based authentication with the Semantic Kernel SDK

> 1. Initialize a `DefaultAzureCredential` object to pickup your app's manged identity:

Reviewer: ```suggestion
1. Initialize a `DefaultAzureCredential` object to pick up your app's managed identity:
```

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="tokenCredential":::

> 1. Build a `Kernel` object including the Azure OpenAI Chat Completion Service, use the perviously created credentials:

Reviewer: ```suggestion
1. Build a `Kernel` object including the Azure OpenAI Chat Completion Service, and use the previously created credentials:
```

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="kernelBuild":::

1. Use the `Kernel` object to invoke prompt completion through Azure OpenAI:

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="invokePrompt":::

## Related content

<!-- Update links once the other docs are done -->

> * [Data Security with App Service]<!-- (link.md) -->

Reviewer: Add link here? Or comment out for now?

Author: This doc is coming next, so no link for now
